jcstress-samples/src/main/java/org/openjdk/jcstress/samples/problems/classic/Classic_02_ProducerConsumerProblem.java

```diff
@@ -0,0 +1,359 @@
+        See https://en.wikipedia.org/wiki/Producer%E2%80%93consumer_problem for more information
+        about the problem and solutions.
+
+        The producer-consumer problem is about transferring items from the producer(s) to the consumer(s)
+        in a thread-safe way. Some solutions support only one producer and one consumer while other solutions
+        don't mind many producers and many consumers.
+     */
+
+    private final static int BUFFER_SIZE = 2;
+
+    /*
+      ----------------------------------------------------------------------------------------------------------
+
+        One general solution uses semaphores to solve the producer-consumer problem.
+     */
+    static class SemaphoresBase {
+        protected final Semaphore fillCount = new Semaphore(BUFFER_SIZE);
+        protected final Semaphore emptyCount = new Semaphore(BUFFER_SIZE);
+
+        public SemaphoresBase() {
+            try {
+                fillCount.acquire(BUFFER_SIZE);
+            } catch (InterruptedException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        public int produce() {
+            try {
+                emptyCount.acquire();
+                int index = putItemIntoBuffer();
+                fillCount.release();
+                return index;
+            } catch (InterruptedException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        public int consume() {
+            try {
+                int count = count();
+                fillCount.acquire();
+                takeItemFromBuffer();
+                emptyCount.release();
+                return count;
+            } catch (InterruptedException e) {
+                throw new IllegalStateException(e);
+            }
+        }
+
+        protected int putItemIntoBuffer() {
+            return count();
+        }
+
+        protected void takeItemFromBuffer() { }
+
+        protected int count() {
+            return fillCount.availablePermits();
+        }
+    }
+
+    /*
+      ----------------------------------------------------------------------------------------------------------
+
+        This solution shows how semaphores can be used to solve the producer-consumer problem
+        for only one producer and only one consumer.
+
+          RESULT      SAMPLES     FREQ      EXPECT  DESCRIPTION
+            true  685.944.832  100,00%  Acceptable  Trivial
+     */
+    @JCStressTest
+    @Outcome(id = "true", expect = ACCEPTABLE, desc = "Trivial")
+    @State
+    public static class OneProducerOneConsumer extends SemaphoresBase {
+        @Actor
+        void producer() {
+            produce();
+            produce();
+        }
+
+        @Actor
+        void consumer() {
+            consume();
+            consume();
+        }
+
+        @Arbiter
+        public void fake(Z_Result r) {
+            r.r1 = true;
```

Comment:
@mmirwaldt , It does not seem to be good to have "all outcomes are acceptable".

Have you considered producing messages with different **values** so you can tell which one was consumed?

For instance: `r1=produce(1); r2=produce(2);` and then `r3=consume(); r4=consume();`.

Then you could verify:
1) "1, 2, 1, 2" is acceptable
2) all others states are forbidden.

The same would apply to the other samples.

WDYT?